rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/shared_functions.cpp

```diff
@@ -0,0 +1,876 @@
+    rmw_free(node_info->subscriber_listener);
+    node_info->subscriber_listener = nullptr;
+  }
+
+  rmw_free(node_info);
+  node->data = nullptr;
+  rmw_node_free(node);
+
+  return RMW_RET_OK;
+}
+
+rmw_guard_condition_t *
+create_guard_condition(const char * implementation_identifier)
+{
+  rmw_guard_condition_t * guard_condition = rmw_guard_condition_allocate();
+  if (!guard_condition) {
+    RMW_SET_ERROR_MSG("failed to allocate guard condition");
+    return NULL;
+  }
+  // Allocate memory for the DDSGuardCondition object.
+  DDSGuardCondition * dds_guard_condition = nullptr;
+  void * buf = rmw_allocate(sizeof(DDSGuardCondition));
+  if (!buf) {
+    RMW_SET_ERROR_MSG("failed to allocate memory");
+    goto fail;
+  }
+  // Use a placement new to construct the DDSGuardCondition in the preallocated buffer.
+  RMW_TRY_PLACEMENT_NEW(dds_guard_condition, buf, goto fail, DDSGuardCondition)
+  buf = nullptr;  // Only free the dds_guard_condition pointer; don't need the buf pointer anymore.
+  guard_condition->implementation_identifier = implementation_identifier;
+  guard_condition->data = dds_guard_condition;
+  return guard_condition;
+fail:
+  if (guard_condition) {
+    rmw_guard_condition_free(guard_condition);
+  }
+  if (buf) {
+    rmw_free(buf);
+  }
+  return NULL;
+}
+
+rmw_ret_t
+destroy_guard_condition(const char * implementation_identifier,
+  rmw_guard_condition_t * guard_condition)
+{
+  if (!guard_condition) {
+    RMW_SET_ERROR_MSG("guard condition handle is null");
+    return RMW_RET_ERROR;
+  }
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    guard condition handle,
+    guard_condition->implementation_identifier, implementation_identifier,
+    return RMW_RET_ERROR)
+
+  auto result = RMW_RET_OK;
+  RMW_TRY_DESTRUCTOR(
+    ((DDSGuardCondition *)guard_condition->data)->~DDSGuardCondition(),
+    DDSGuardCondition, result = RMW_RET_ERROR)
+  rmw_guard_condition_free(guard_condition);
+  return result;
+}
+
+rmw_ret_t
+trigger_guard_condition(const char * implementation_identifier,
+  const rmw_guard_condition_t * guard_condition_handle)
+{
+  if (!guard_condition_handle) {
+    RMW_SET_ERROR_MSG("guard condition handle is null");
+    return RMW_RET_ERROR;
+  }
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    guard condition handle,
+    guard_condition_handle->implementation_identifier, implementation_identifier,
+    return RMW_RET_ERROR)
+
+  DDSGuardCondition * guard_condition =
+    static_cast<DDSGuardCondition *>(guard_condition_handle->data);
+  if (!guard_condition) {
+    RMW_SET_ERROR_MSG("guard condition is null");
+    return RMW_RET_ERROR;
+  }
+  DDS_ReturnCode_t status = guard_condition->set_trigger_value(DDS_BOOLEAN_TRUE);
+  if (status != DDS_RETCODE_OK) {
+    RMW_SET_ERROR_MSG("failed to set trigger value");
+    return RMW_RET_ERROR;
+  }
+  return RMW_RET_OK;
+}
+
+template<typename SubscriberInfo, typename ServiceInfo, typename ClientInfo>
```

Comment:
Does this have to be a template or can these info struct be unified to be the same across both implementations?